main.go

```diff
@@ -8,141 +8,131 @@ import (
-	if err != nil {
-		panic(err)
+	failed := false
+	for i, v := range targets {
+		ua, err := validateUrl(v)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			failed = true
+			continue
+		}
+		urls[i] = ua
 	}
-	ub, err := url.Parse(targets[1])
-	if err != nil {
-		panic(err)
+	if failed {
+		os.Exit(1)
 	}
 
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-		resp_a := make(chan *http.Response)
-		resp_b := make(chan *http.Response)
+		// Allocate a channel for the HTTP responses.
+		responses := make(chan *http.Response, len(targets))
 
-		fail_a := make(chan struct{})
-		fail_b := make(chan struct{})
-		fail := make(chan struct{})
+		// Allocate a cancellation channel for each target's HTTP request.
+		cancels := make([]chan struct{}, len(targets))
 
-		cancel_a := make(chan struct{})
-		cancel_b := make(chan struct{})
+		// Allocate a failure channel for each target's HTTP request.
+		fails := make([]chan struct{}, len(targets))
 
 		r.RequestURI = ""
 		r.URL.Scheme = "http"
 
-		go func() {
-			req_a := *r
-			req_a.URL.Host = ua.Host
-			req_a.URL, _ = url.Parse(r.URL.String())
-			req_a.Cancel = cancel_a
-
-			rt := &http.Transport{DisableKeepAlives: true}
-			resp, err := rt.RoundTrip(&req_a)
-			if err != nil {
-				log.Printf("target A failed: %s", err)
-				close(fail_a)
-			} else if resp.StatusCode >= 500 || resp.StatusCode == 408 {
-				log.Printf("target A unsatisfying status: %d", resp.StatusCode)
-				close(fail_a)
-			} else {
-				resp_a <- resp
-			}
-		}()
+		// Create and send out an HTTP request to each target.
+		for i := range targets {
+			req := *r
+			req.URL.Host = urls[i].Host
+			req.URL, _ = url.Parse(r.URL.String())
+			cancels[i] = make(chan struct{})
+			fails[i] = make(chan struct{})
+			req.Cancel = cancels[i]
+			fail := fails[i]
+
+			go func() {
+				rt := &http.Transport{DisableKeepAlives: true}
+				res, err := rt.RoundTrip(&req)
+				if err != nil {
+					log.Printf("request failed: %s\n", err)
+					close(fail)
+				} else if res.StatusCode >= 500 || res.StatusCode == 408 {
+					log.Printf("target (%s) unsatisfying status: %d", req.URL, res.StatusCode)
+					close(fail)
+				} else {
+					log.Printf("target (%s) responded with %d", req.URL, res.StatusCode)
+					responses <- res
+				}
+			}()
+		}
 
+		// Listen to all requests' failure channels; issue a signal when all requests fail.
+		failed := make(chan struct{})
 		go func() {
-			req_b := *r
-			req_b.URL, _ = url.Parse(r.URL.String())
-			req_b.URL.Host = ub.Host
-			req_b.Cancel = cancel_b
-
-			rt := &http.Transport{DisableKeepAlives: true}
-			resp, err := rt.RoundTrip(&req_b)
-			if err != nil {
-				log.Printf("target B failed: %s", err)
-				close(fail_b)
-			} else if resp.StatusCode >= 500 || resp.StatusCode == 408 {
-				log.Printf("target B unsatisfying status: %d", resp.StatusCode)
-				close(fail_b)
-			} else {
-				resp_b <- resp
+			for i := range fails {
+				<-fails[i]
```

Comment:
this is back to the point where I don't think this goroutine is ever gonna exit. where do *all* the fails get closed?